libtenzir/builtins/operators/buffer.cpp

```diff
@@ -0,0 +1,408 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/detail/enum.hpp>
+#include <tenzir/detail/weak_run_delayed.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/tql2/plugin.hpp>
+#include <tenzir/uuid.hpp>
+
+#include <caf/typed_event_based_actor.hpp>
+
+#include <queue>
+
+namespace tenzir::plugins::buffer {
+
+TENZIR_ENUM(buffer_policy, block, drop);
+
+namespace {
+
+using buffer_actor = caf::typed_actor<
+  // Write events into the buffer.
+  auto(atom::write, table_slice events)->caf::result<void>,
+  // Read events from the buffer.
+  auto(atom::read)->caf::result<table_slice>>;
+
+struct buffer_state {
+  static constexpr auto name = "buffer";
+
+  buffer_actor::pointer self = {};
+  located<uint64_t> capacity = {};
+  buffer_policy policy = {};
+  metric_handler metrics_handler = {};
+  shared_diagnostic_handler diagnostics_handler = {};
+
+  uint64_t buffer_size = {};
+  std::queue<table_slice> buffer = {};
+  caf::typed_response_promise<table_slice> read_rp = {};
+
+  table_slice blocked_events = {};
+  caf::typed_response_promise<void> write_rp = {};
+
+  uint64_t num_dropped = {};
+
+  ~buffer_state() noexcept {
+    emit_metrics();
+    if (read_rp.pending()) {
+      read_rp.deliver(table_slice{});
+    }
+  }
+
+  auto write(table_slice events) -> caf::result<void> {
+    if (read_rp.pending()) {
+      read_rp.deliver(std::move(events));
+      return {};
+    }
+    if (buffer_size + events.rows() > capacity.inner) {
+      auto [lhs, rhs] = split(events, capacity.inner - buffer_size);
+      buffer_size += lhs.rows();
+      buffer.push(std::move(lhs));
+      TENZIR_ASSERT(rhs.rows() > 0);
+      switch (policy) {
+        case buffer_policy::drop: {
+          num_dropped += rhs.rows();
+          diagnostic::warning("buffer exceeded capacity and dropped events")
+            .primary(capacity.source)
```

Comment:
Absolute nitpick: This places the diagnostic on the buffer size (better would be the operator name or even the whole operator invocation, including arguments).

Reply:
Is there a way to get the entire operator invocation easily?

Reply:
Would need to add to add a `get_location` helper to `invocation`. If you just want to use the, you can use `inv.self.get_location()`.

Reply:
I think I'll leave this as-is for now. Pointing to the capacity is at least directly pointing to the number the user likely should increase.

Reply:
I left a TODO comment in 28ae3c2e65d